tree.go

```diff
@@ -467,6 +453,38 @@ func (n *InternalNode) Delete(key []byte, resolver NodeResolverFn) error {
 // with HashedNode. It also sends the current node on the flush channel.
 func (n *InternalNode) Flush(flush NodeFlushFn) {
 	n.Commit()
+
+	if n.depth <= 0 {
```

Comment:
Here we're doing what I described in the [go-ethereum PR](https://github.com/gballet/go-ethereum/pull/147/#discussion_r1047880009) (actually, can be a good opportunity to have both side by side).

The TL;DR is that if we're flushing the root node, this will trigger a recursive call to all the children which will do `Commit()` work. This is quite heavy as we know. What we do here, is doing this in parallel since each child is independent from others.

To avoid an explosion of goroutines which will hurt performance because of scheduling, we do this in `batches := runtime.NumCPU()`. I've tested with a range and this seems to be a good choice.

Reply:
The total allocations (number and space) has gone down in this PR by a dramatic factor, so my gut feeling says it shouldn't be a problem (actually, it should be even better).

If in the future we focus on the conversion process later, we can double-check our intuition. If, for whatever reason, it's a problem, we have some flag to make it optional to parallelize the work, so we can have the best of both worlds. 

Even without memory pressure situations, it might make sense if the conversion process is already splitting up work in goroutines at that (high) level. So, it can be a good idea already to remember.